fastly/pool.go

```diff
@@ -0,0 +1,293 @@
+		return nil, ErrMissingService
+	}
+
+	if i.Version == 0 {
+		return nil, ErrMissingVersion
+	}
+
+	if i.Name == "" {
+		return nil, ErrMissingName
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/pool", i.Service, i.Version)
+	resp, err := c.PostForm(path, i, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var p *Pool
+	if err := decodeJSON(&p, resp.Body); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+// GetPoolInput is used as input to the GetPool function.
+type GetPoolInput struct {
+	// Service is the ID of the service (required).
+	Service string
+
+	// Version is the specific configuration version (required).
+	Version int
+
+	// Name is the name of the pool of interest (required).
+	Name string
+}
+
+// GetPool gets a single pool for a particular service and version.
+func (c *Client) GetPool(i *GetPoolInput) (*Pool, error) {
+	if i.Service == "" {
+		return nil, ErrMissingService
+	}
+
+	if i.Version == 0 {
+		return nil, ErrMissingVersion
+	}
+
+	if i.Name == "" {
+		return nil, ErrMissingName
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/pool/%s", i.Service, i.Version, url.PathEscape(i.Name))
+	resp, err := c.Get(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var p *Pool
+	if err := decodeJSON(&p, resp.Body); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+// UpdatePoolInput is used as input to the UpdatePool function.
+type UpdatePoolInput struct {
+	// Service is the ID of the service. Version is the specific configuration
+	// version. Both fields are required.
+	Service string
+	Version int
+
+	// Name is the name of the pool to update (required).
+	Name string
+
+	NewName          string       `form:"name,omitempty"`
+	Comment          string       `form:"comment,omitempty"`
+	Shield           string       `form:"shield,omitempty"`
+	RequestCondition string       `form:"request_condition,omitempty"`
+	MaxConnDefault   uint         `form:"max_conn_default,omitempty"`
+	ConnectTimeout   uint         `form:"connect_timeout,omitempty"`
+	FirstByteTimeout uint         `form:"first_byte_timeout,omitempty"`
+	Quorum           uint         `form:"quorum,omitempty"`
+	UseTLS           *Compatibool `form:"use_tls,omitempty"`
+	TLSCACert        string       `form:"tls_ca_cert,omitempty"`
+	TLSCiphers       string       `form:"tls_ciphers,omitempty"`
+	TLSClientKey     string       `form:"tls_client_key,omitempty"`
+	TLSClientCert    string       `form:"tls_client_cert,omitempty"`
+	TLSSNIHostname   string       `form:"tls_sni_hostname,omitempty"`
+	TLSCheckCert     bool         `form:"tls_check_cert,omitempty"`
+	TLSCertHostname  string       `form:"tls_cert_hostname,omitempty"`
+	MinTLSVersion    string       `form:"min_tls_version,omitempty"`
+	MaxTLSVersion    string       `form:"max_tls_version,omitempty"`
+	Healthcheck      string       `form:"healthcheck,omitempty"`
+	Type             PoolType     `form:"type,omitempty"`
+	OverrideHost     string       `form:"override_host,omitempty"`
+}
```

Comment:
Not an issue with your implementation, as we have this problem throughout the project. However there is a problem with using scalar types directly in `UpdateFooInput` input structs, and we intend to move all other structs to this pattern over this year.

The problem is the combination of Go's type zero values, JSON marshalling with the omitempty struct field tag, and a users intent to actually set a zero value. This blog post does a much better job at explaining the problem and well worth a read: https://willnorris.com/2014/05/go-rest-apis-and-pointers/

So in your example, if `UpdatePoolInput.TLSCheckCert` is currently `true` and the user attempts to update the value to `false`, omitempty considers this value empty and won't send the updated value.

As the post suggests, the more robust solution would be to change the struct field values to all be pointers instead of the underlying type value. However (whilst we think this is the correct approach) it adds some pain for the library users to convert their inputs to pointers and do nil pointer checks when consuming etc.

Therefore, what are your thoughts on changing all these values to pointers?

Reply:
I'm definitely in favor of using the pointers in the struct fields. It's true that it makes it a bit more cumbersome for users to set the fields, though a set of helpers (cf. [goprotobuf](https://github.com/golang/protobuf/blob/ed6926b37a637426117ccab59282c3839528a700/proto/lib.go#L457-L461) and [aws](https://github.com/aws/aws-sdk-go/blob/v1.28.5/aws/convert_types.go#L182-L185)) helps. 

I suspect the response structs the lib would want to also use pointers for fields (if nothing else, to keep consistency between inputs and outputs), though it may not be strictly necessary since it seems like the Fastly API always returns all fields in the response struct.

Without using pointers for field structs, there are definitely cases where resources won't be able to be modified, for the Pool resource alone one won't be able to:

* Disable quorom once enabled
* Remove TLSCaCert
* Remove TLSCiphers
* Remove TLSClientKey
* Remove TLSClientCert
* Remove TLSSNIHostname
* Disable Healthcheck
* Remove Comment
* Remove OverrideHost
* Disable TLSCheckCert

I think that list alone makes a compelling case for using pointer fields. Updating the library throughout is undoubtedly a breaking change, but if you want to pave the way with these new resources I'm all for it.

Reply:
> I think that list alone makes a compelling case for using pointer fields. Updating the library throughout is undoubtedly a breaking change, but if you want to pave the way with these new resources I'm all for it.

Apologies for the confusion, yes I'd like to start this process for these new resources. You don't need to worry about updating all of the existing ones :)

Reply:
👍 Let's do it! I'll make the changes.